### PR TITLE
Fix Javascript injection & ReDoS vulnerabilities

### PR DESCRIPTION
**Pull request by scottinet**
# Description

Fix harmful uses of JS keywords in requests made to Kuzzle, due to a lack of distrust about external data.

Example: an index named `constructor` is stored as a key, and then used like this: `this.someCacheObject[index]`, effectively invoking Object.constructor, which is a Javascript reserved keyword.

This issue is solved by:
- storing our RAM cache in maps
- adding a new utility module made to safely check or access to object properties (with 50x to 300x better performances than the equivalent lodash methods)

### CODE REVIEW

File: lib/api/kuzzle.js [hunk 1]
```diff
@@ -314,7 +314,9 @@ function dumpAndExit(kuzzle, suffix) {
     runShutdown(kuzzle);
   }
 
-  kuzzle.janitor.dump(suffix).finally(() => runShutdown(kuzzle));
+  kuzzle.janitor.dump(suffix)
+    .catch(() => {}) // do nothing, prevents unhandled rejections
```
Comment (kuzzle) [inline lib/api/kuzzle.js]: CRITICAL Add a nested comment explaining why this function is empty or complete the implementation. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1186)
MAJOR Either remove or fill this block of code. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AEmptyBlock)

File: lib/api/core/hotelClerk.js [hunk 2]
```diff
@@ -134,13 +136,13 @@ class HotelClerk {
...
       roomId = request.input.body.roomId,
-      room = this.rooms[roomId];
+      room = this.rooms.get(roomId);
 
     if (room === undefined) {
       realtimeError.throw('room_not_found', roomId);
     }
 
-    return {count: this.rooms[roomId].customers.size};
+    return {count: room.customers.size};
```
Comment (kuzzle) [inline lib/api/core/hotelClerk.js]: MAJOR TypeError can be thrown as "room" might be null or undefined here. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS2259)

File: lib/api/core/plugins/pluginsManager.js [hunk 3]
```diff
@@ -918,7 +915,7 @@ class PluginsManager {
     // @todo the _.isNil test need to be removed as soon as the
```
Comment (kuzzle) [inline lib/api/core/plugins/pluginsManager.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

**Comment by kuzzle** [comment]
SonarQube analysis reported 5 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 2 major
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [collectionController.js#L434](https://github.com/kuzzleio/kuzzle/blob/13c15404973d26c3f98127265643eb0c884106c8/lib/api/controllers/collectionController.js#L434): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by kuzzle** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [collectionController.js#L434](https://github.com/kuzzleio/kuzzle/blob/2f31847d1984ef94f5dcd98c6124bd623dee04d1/lib/api/controllers/collectionController.js#L434): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: lib/api/core/hotelClerk.js (hunk 2 above)
Comment (scottinet) [reply lib/api/core/hotelClerk.js]: No, it cannot you moron.

File: lib/api/core/hotelClerk.js [hunk 4]
```diff
@@ -134,13 +136,13 @@ class HotelClerk {
...
-      room = this.rooms[roomId];
+      room = this.rooms.get(roomId);
 
-    if (room === undefined) {
+    if (room === undefined || room === null) {
       realtimeError.throw('room_not_found', roomId);
     }
 
-    return {count: this.rooms[roomId].customers.size};
+    return {count: room.customers.size};
```
Comment (kuzzle) [inline lib/api/core/hotelClerk.js]: MAJOR TypeError can be thrown as "room" might be null or undefined here. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS2259)

**Comment by kuzzle** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [collectionController.js#L434](https://github.com/kuzzleio/kuzzle/blob/e07652157ddd869d92d59319d80f7b121dd8db14/lib/api/controllers/collectionController.js#L434): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by kuzzle** [comment]
SonarQube analysis reported 2 issues
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [collectionController.js#L434](https://github.com/kuzzleio/kuzzle/blob/50b4a63b23c9c5fb31feb1cace96dfc3bdea4c16/lib/api/controllers/collectionController.js#L434): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Pull request merged**